runtime/criusupport/criusupport.cpp

```diff
@@ -803,16 +824,23 @@ Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl(JNIEnv *env,
 
 		acquireSafeOrExcusiveVMAccess(currentThread, vmFuncs, safePoint);
 
-		for (UDATA i = 0; !checkIfSafeToCheckpoint(currentThread) && retryPermitted; i++) {
+		for (UDATA i = 0; (0 != (notSafeToCheckpoint = checkIfSafeToCheckpoint(currentThread))) && retryPermitted; i++) {
```

Comment:
That's a lot of logic in the loop condition.  Can the `checkIfSafeToCheckpoint` be moved so it's done either always in the loop or out?

retryPermitted is constant for the duration of the function - can you make it `const`?

Actually, it might be to make a local that holds "maxRetries" and is initialized to `vm->checkpointState.maxRetryForNotCheckpointSafe`.

Then something like:
```
notSafeToCheckpoint = checkIfSafeToCheckpoint(currentThread);
for (int i = 0; (0 != notSafeToCheckpoint ) && (i <= maxRetries); i++) {
     /// existing loop code
     notSafeToCheckpoint = checkIfSafeToCheckpoint(currentThread);
}
```

Also the extra check for `i == vm->checkpointState.maxRetryForNotCheckpointSafe` in the loop seems to complicate the loop.  It ca be hoisted out and only checked after the loop:
```
notSafeToCheckpoint = checkIfSafeToCheckpoint(currentThread);
for (int i = 0; (0 != notSafeToCheckpoint ) && (i <= maxRetries); i++) {
     /// existing loop code
     notSafeToCheckpoint = checkIfSafeToCheckpoint(currentThread);
}
// either checkpointSafe or past maxRetries at this point
if ((J9VM_DELAYCHECKPOINT_NOTCHECKPOINTSAFE == notSafeToCheckpoint)
|| ((J9VM_DELAYCHECKPOINT_CLINIT == notSafeToCheckpoint) 
     && J9_ARE_ALL_BITS_SET(vm->checkpointState.flags, J9VM_CRIU_IS_THROW_ON_DELAYED_CHECKPOINT_ENABLED))
 ) {
 	currentExceptionClass = vm->checkpointState.criuJVMCheckpointExceptionClass;
 	systemReturnCode = vm->checkpointState.maxRetryForNotCheckpointSafe;
 	nlsMsgFormat = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE, J9NLS_JCL_CRIU_MAX_RETRY_FOR_NOTCHECKPOINTSAFE_REACHED, NULL);
 	goto closeWorkDirFD;
 } else {
 	Trc_CRIU_checkpointJVMImpl_checkpointWithActiveCLinit(currentThread);
 	break;
 }
```
Without this refactoring, in the case of `J9VM_DELAYCHECKPOINT_CLINIT` when not throwing, you won't `acquireSafeOrExcusiveVMAccess` as it breaks out of the loop.